packages/@aws-cdk/core/lib/arn.ts

```diff
@@ -76,10 +76,13 @@ export class Arn {
    * the 'scope' is attached to. If all ARN pieces are supplied, the supplied scope
    * can be 'undefined'.
    */
-  public static format(components: ArnComponents, stack: Stack): string {
-    const partition = components.partition !== undefined ? components.partition : stack.partition;
-    const region = components.region !== undefined ? components.region : stack.region;
-    const account = components.account !== undefined ? components.account : stack.account;
+  public static format(components: ArnComponents, stack?: Stack): string {
+    if (stack === undefined && (components.partition === undefined || components.region === undefined || components.account === undefined)) {
+      throw new Error('Stack must be defined if any individual component is undefined');
```

Comment:
let's list the components out so this error message becomes more of a help message in guiding users that run into it